### PR TITLE
Payment channel on Polkadot

### PR DESCRIPTION
**Pull request by leandro-ro**
On the basis of the Ethereum payment channel code.
I tried to make as minimal changes as possible:

Client/client.go:
- Replaced `swallet` with `dotwallet`
- Replaced `common.Address` with `wallet.Address`
- Use `dotchannel.Asset` as asset
- Replaced `CreateContractBackend` with `dot.NewAPI`
- Create `pallet` (using dot api)
    - Derive adjudicator & funder from there (no contract deployment)

Client/channel.go
- `DotToPlanck` (instead of `EthToWei`)

Client/handle.go
- `clientIdx`: Funding now expected to be equal on both sides 
    - Note: Deadlock if funding from one side is zero. We might need to investigate what’s going on there.

Client/util.go
- `WireAddress` and `WalletAddress` now equally derived
    - For the sake of consistency, still two separate methods

util.go
- Use `dotwallet`
- Balance logger adapted to dot

main.go
- `chainID` now `NetworkID`
- `blockQueryDepth`
- No key for deployer necessary

### CODE REVIEW

File: payment-channel-dot/client/client.go [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+	networkId dot.NetworkID, // networkId is the identifier of the blockchain.
+	queryDepth types.BlockNumber, // queryDepth is the number of blocks being evaluated when looking for events.
+) (*PaymentClient, error) {
+	// Connect to backend.
+	api, err := dot.NewAPI(nodeURL, networkId)
+	if err != nil {
+		panic(err)
+	}
+
+	// Create Perun pallet and generate funder + adjudicator from it.
```
Comment (matthiasgeihs) [inline payment-channel-dot/client/client.go:59]: Technically, you don't create a pallet here. The pallet is already running as part of the node. You just establish a connection to the pallet and then bootstrap the funder and adjudicator from that.

**Comment by matthiasgeihs** [comment]
> Client/handle.go
> 
> * `clientIdx`: Funding now expected to be equal on both sides
>   
>   * Note: Deadlock if funding from one side is zero. We might need to investigate what’s going on there.


Can you open an issue in the Polkadot backend repo? @leandro-ro

**Review by matthiasgeihs — COMMENTED** [review]
looks quite good 👍 

note for next time: would have been easier to spot the differences to the Ethereum version if you start with a commit of a copy of the Ethereum version and then apply the changes on top.

**Comment by leandro-ro** [comment]
> > Client/handle.go
> > 
> > * `clientIdx`: Funding now expected to be equal on both sides
> >   
> >   * Note: Deadlock if funding from one side is zero. We might need to investigate what’s going on there.
> 
> Can you open an issue in the Polkadot backend repo? @leandro-ro

I'm not quite sure if this is even a bug. In polkadot the [existential deposit](https://support.polkadot.network/support/solutions/articles/65000168651-what-is-the-existential-deposit-)  is 1 DOT (or 0.0000333333 KSM). But at least an error message (& not running into a deadlock) would be nice of course.

File: payment-channel-dot/client/client.go (hunk 1 above)
Comment (matthiasgeihs) [reply payment-channel-dot/client/client.go]: The intention was that you adapt the comment, but not super important now.

**Comment by matthiasgeihs** [comment]
> > > Client/handle.go
> > > 
> > > * `clientIdx`: Funding now expected to be equal on both sides
> > >   
> > >   * Note: Deadlock if funding from one side is zero. We might need to investigate what’s going on there.
> > 
> > 
> > Can you open an issue in the Polkadot backend repo? @leandro-ro
> 
> I'm not quite sure if this is even a bug. In polkadot the [existential deposit](https://support.polkadot.network/support/solutions/articles/65000168651-what-is-the-existential-deposit-) is 1 DOT (or 0.0000333333 KSM). But at least an error message (& not running into a deadlock) would be nice of course.

Yes, it might have to do something with that. But it should definitely still be possible to open channels where one side does not deposit funds.

**Review by matthiasgeihs — APPROVED** [review]

**Pull request merged**